Cataloger ERROR: value too long for type character varying(255)

I think this is happening writing very long Institution Names.

Sample SCOAP record that demonstrates this problem:
http://repo.scoap3.org/record/9107/files/main.xml

```
in process about to evaluate: //datafield[@tag='700']/subfield[(@code='v')]
keyParts0: //datafield[@tag='700']/subfield[(@code='v')]
Post eval num hits: 2899

[[SNIP a few thousand lines of dumped Institution Names]]]

doc2p: fname: export/xm?ln=en
IDHits size: 2899
[ERROR] [03/11/2015 13:39:12.418] [application-akka.actor.default-dispatcher-6] [akka://application/user/cataloger] ERROR: value too long for type character varying(255)
org.postgresql.util.PSQLException: ERROR: value too long for type character varying(255)
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2270)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1998)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:255)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:570)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:420)
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeUpdate(AbstractJdbc2Statement.java:366)
	at com.jolbox.bonecp.PreparedStatementHandle.executeUpdate(PreparedStatementHandle.java:205)
	at anorm.Sql$class.execute1(Anorm.scala:321)
	at anorm.SimpleSql.execute1(Anorm.scala:190)
	at anorm.Sql$class.executeInsert(Anorm.scala:351)
	at anorm.SimpleSql.executeInsert(Anorm.scala:190)
	at models.Topic$$anonfun$create$1.apply(Topic.scala:124)
	at models.Topic$$anonfun$create$1.apply(Topic.scala:118)
	at play.api.db.DBApi$class.withConnection(DB.scala:81)
	at play.api.db.BoneCPApi.withConnection(DB.scala:267)
	at play.api.db.DB$$anonfun$withConnection$3.apply(DB.scala:162)
	at play.api.db.DB$$anonfun$withConnection$3.apply(DB.scala:162)
	at scala.Option.map(Option.scala:145)
	at play.api.db.DB$.withConnection(DB.scala:162)
	at models.Topic$.create(Topic.scala:118)
	at workers.Cataloger.createTopic(Cataloger.scala:177)
	at workers.Cataloger$$anonfun$topics$1$$anonfun$apply$1.apply(Cataloger.scala:82)
	at workers.Cataloger$$anonfun$topics$1$$anonfun$apply$1.apply(Cataloger.scala:78)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at workers.Cataloger$$anonfun$topics$1.apply(Cataloger.scala:78)
	at workers.Cataloger$$anonfun$topics$1.apply(Cataloger.scala:70)
	at scala.collection.immutable.List.map(List.scala:272)
	at workers.Cataloger.topics(Cataloger.scala:70)
	at workers.Cataloger$$anonfun$catalog$2.apply(Cataloger.scala:306)
	at workers.Cataloger$$anonfun$catalog$2.apply(Cataloger.scala:306)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at workers.Cataloger$.catalog(Cataloger.scala:306)
	at workers.CatalogWorker$$anonfun$receive$1.applyOrElse(Cataloger.scala:33)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
	at workers.CatalogWorker.aroundReceive(Cataloger.scala:31)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
	at akka.actor.ActorCell.invoke(ActorCell.scala:487)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
	at akka.dispatch.Mailbox.run(Mailbox.scala:220)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

--- a/app/controllers/Application.scala
+++ b/app/controllers/Application.scala
@@ -874,6 +874,20 @@ object Application extends Controller {
     ).getOrElse(NotFound(views.html.static.trouble("No such hold: " + id)))
   }
 
+  def pickBrowse(id: Int, page: Int) = Action { implicit request =>
+    Subscriber.findById(id).map( sub =>
+      Ok(views.html.topic_pick.browse(sub.id, sub.picks(page), page, sub.pickCount))
+    ).getOrElse(NotFound(views.html.static.trouble("No such subscriber: " + id)))
+  }
+
+  def resolvePick(id: Int, accept: Boolean) = Action { implicit request =>
+    TopicPick.findById(id).map( pick => {
+      conveyor ! (pick, accept)
+      Redirect(routes.Application.pickBrowse(1, 0))
+    }
+    ).getOrElse(NotFound(views.html.static.trouble("No such topic pick: " + id)))
+  }
+
   val modelForm = Form(
     single(
       "model" -> nonEmptyText

--- a/app/models/Agent.scala
+++ b/app/models/Agent.scala
@@ -1,0 +1,71 @@
+/**
+  * Copyright (c) 2015 MIT Libraries
+  * Licensed under: http://www.apache.org/licenses/LICENSE-2.0
+  */
+package models
+
+import play.api.db.DB
+import play.api._
+import play.api.Play.current
+
+import anorm.SqlParser._
+import anorm.~
+import anorm.SQL
+
+/** An agent is a software service that can perform operations for
+  * benefit of subscribers, such as suggesting topics to subscribe to,
+  * automatically subscribing or unsubscribing, etc
+  *
+  * @author richardrodgers
+  */
+
+case class Agent(id: Int,
+                 tag: String,
+                 label: String,
+                 description: String,
+                 code: String,
+                 params: String,
+                 icon: Option[String])
+
+object Agent {
+
+  val agent = {
+    get[Int]("id") ~ get[String]("tag") ~ get[String]("label") ~ get[String]("description") ~
+    get[String]("code") ~ get[String]("params") ~ get[Option[String]]("icon") map {
+      case id ~ tag ~ label ~ description ~ code ~ params ~ icon => Agent(id, tag, label, description, code, params, icon)
+    }
+  }
+
+  def create(tag: String, label: String, description: String, code: String, params: String, icon: Option[String]) = {
+		DB.withConnection { implicit c =>
+			SQL("insert into agent (tag, label, description, code, params, icon) values ({tag}, {label}, {description}, {code}, {params}, {icon})")
+      .on('tag -> tag, 'label -> label, 'description -> description, 'code -> code, 'params -> params, 'icon -> icon).executeInsert()
+		}
+  }
+
+  def make(tag: String, label: String, description: String, code: String, params: String, icon: Option[String]) = {
+    findById(create(tag, label, description, code, params, icon).get.toInt).get
+  }
+
+  def all: List[Agent] = {
+    DB.withConnection { implicit c =>
+      SQL("select * from agent").as(agent *)
+    }
+  }
+
+  def findById(id: Int): Option[Agent] = {
+    DB.withConnection { implicit c =>
+      SQL("select * from agent where id = {id}").on('id -> id).as(agent.singleOpt)
+    }
+  }
+
+  def findByTag(tag: String): Option[Agent] = {
+    DB.withConnection { implicit c =>
+      SQL("select * from agent where tag = {tag}").on('tag -> tag).as(agent.singleOpt)
+    }
+  }
+
+  def mapView: Map[String, String] = {
+    all map (cp => cp.id.toString -> cp.tag) toMap
+  }
+}

--- a/app/models/Subscriber.scala
+++ b/app/models/Subscriber.scala
@@ -130,6 +130,33 @@ case class Subscriber(id: Int,  // DB key
     }
   }
 
+  def pickCount = {
+    DB.withConnection { implicit c =>
+      val count = SQL("select count(*) as c from topic_pick where subscriber_id = {id}").on('id -> id).apply.head
+      count[Long]("c")
+    }
+  }
+
+  def picked(topicId: Int): Option[TopicPick] = {
+    DB.withConnection { implicit c =>
+      SQL("select * from topic_pick where topic_id = {topic_id} and subscriber_id = {id}")
+      .on('topic_id -> topicId, 'id -> id).as(TopicPick.pick.singleOpt)
+    }
+  }
+
+  def picks(page: Int) = {
+    val offset = page * 10
+    DB.withConnection { implicit c =>
+      SQL(
+        """
+          select * from topic_pick where subscriber_id = {sub_id}
+          order by created desc
+          limit 10 offset {offset}
+        """
+      ).on('sub_id -> id, 'offset -> offset).as(TopicPick.pick *)
+    }
+  }
+
   def channels: List[Channel] = {
     DB.withConnection { implicit c =>
       SQL("select * from channel where subscriber_id = {id}").on('id -> id).as(Channel.channel *)

--- a/app/models/TopicPick.scala
+++ b/app/models/TopicPick.scala
@@ -1,0 +1,78 @@
+/**
+  * Copyright (c) 2015 MIT Libraries
+  * Licensed under: http://www.apache.org/licenses/LICENSE-2.0
+  */
+package models
+
+import java.util.Date
+
+import play.api.db.DB
+import play.api._
+import play.api.Play.current
+
+import anorm.SqlParser._
+import anorm.~
+import anorm.SQL
+import anorm.Row
+
+/** TopicPick is a topic suggestion (potential subscription) for a subscriber
+  * Following a review, the pick is resolved to a subscription, or discarded
+  *
+  * @author richardrodgers
+  */
+
+case class TopicPick(id: Int,  // DB key
+                     subscriberId: Int,   // DB key of subscriber
+                     topicId: Int,   // DB key of selected topic
+                     agentId: Int, // DB key of agent that created it
+                     created: Date,  // when pick created
+                     resolved: Date) {    // when pick resolved
+
+  def topic = {
+    DB.withConnection { implicit c =>
+      SQL("select * from topic where id = {topic_id}").on('topic_id -> topicId).as(Topic.topic.singleOpt).get
+    }
+  }
+
+  def resolve(accept: Boolean) = {
+    // not currently remembering state, so just delete
+    DB.withConnection { implicit c =>
+      SQL("delete from topic_pick where id = {id}").on('id -> id).executeUpdate()
+    }
+  }
+}
+
+object TopicPick {
+
+  val pick = {
+    get[Int]("id") ~ get[Int]("subscriber_id") ~ get[Int]("topic_id") ~ get[Int]("agent_id") ~
+    get[Date]("created") ~ get[Date]("resolved") map {
+      case id ~ subscriberId ~ topicId ~ agentId ~ created ~ resolved =>
+        TopicPick(id, subscriberId, topicId, agentId, created, resolved)
+    }
+  }
+
+  def picked(topicId: Int, subscriberId: Int) = {
+    DB.withConnection { implicit c =>
+      SQL("select * from topic_pick where topic_id = {topic_id} and subscriber_id = {subscriber_id}")
+      .on('topic_id -> topicId, 'subscriber_id -> subscriberId).as(pick.singleOpt).isDefined
+    }
+  }
+
+  def findById(id: Int): Option[TopicPick] = {
+     DB.withConnection { implicit c =>
+       SQL("select * from topic_pick where id = {id}").on('id -> id).as(pick.singleOpt)
+     }
+   }
+
+  def create(subscriberId: Int, topicId: Int, agentId: Int) = {
+    DB.withConnection { implicit c =>
+      SQL("insert into topic_pick (subscriber_id, topic_id, agent_id, created, resolved) values ({subscriber_id}, {topic_id}, {agent_id}, {created}, {resolved})")
+      .on('subscriber_id -> subscriberId, 'topic_id -> topicId, 'agent_id -> agentId, 'created -> new Date, 'resolved -> new Date).executeInsert()
+    }
+  }
+
+  def make(subscriberId: Int, topicId: Int, agentId: Int) = {
+    findById(create(subscriberId, topicId, agentId).get.toInt).get
+  }
+}

--- a/app/views/subscriber/dashboard.scala.html
+++ b/app/views/subscriber/dashboard.scala.html
@@ -62,7 +62,13 @@
       </ul>
 
       <h4>Suggested Subscriptions</h4>
-      <ul class="list-group"><li class="list-group-item">Coming (much) Later</li></ul>
+      <ul class="list-group">
+        <li class="list-group-item">
+          <a href="@routes.Application.pickBrowse(1)">Total to review:
+          <span class="badge">@Subscriber.findById(1).get.pickCount</span>
+          </a>
+        </li>
+      </ul>
 
       <h4>Destinations</h4>
       <ul class="list-group">

--- a/app/views/topic/show.scala.html
+++ b/app/views/topic/show.scala.html
@@ -23,6 +23,9 @@
            @if(sub.get.hasInterest(topic.scheme_id)) {
              <span><a rel="tooltip" title="Subscribe to this topic" href="@routes.Application.topicSubscribe(topic.id)" class="btn btn-primary btn-large">Subscribe <span class="badge">@sub.get.newItemCountFor(topic.id)</span> &raquo;</a></span>
            }
+           @if(sub.get.picked(topic.id).isDefined) {
+             <span><a rel="tooltip" title="Forget this topic" href="@routes.Application.resolvePick(sub.get.picked(topic.id).get.id, false)" class="btn btn-primary btn-large pull-right">Meh &raquo;</a></span>
+           }
          }
        }
     </div>

--- a/app/views/topic_pick/browse.scala.html
+++ b/app/views/topic_pick/browse.scala.html
@@ -1,0 +1,17 @@
+@*****************************************************************************
+* Entry page for topic pick browsing (Review)                                *
+* Copyright (c) 2015 MIT Libraries                                           *
+*****************************************************************************@
+@(id: Int, picks: List[TopicPick], page: Int, total: Long)@*(implicit hubContext: HubContext)*@
+@layout.main("Topic Picks Browse - TopicHub") {
+  <div class="page-header">
+    <h2>Review Topic Picks</h2>
+    @tags.summary(page, 10, total, picks.length)
+  </div>
+  <ul>
+    @picks.map { pick =>
+      <li><a href="@routes.Application.topic(pick.topicId)">@pick.topic.tag</a> @pick.topic.name</li>
+    }
+  </ul>
+  @tags.pagination(page, 10, total, routes.Application.pickBrowse(id).toString, picks.length)
+}

--- a/app/workers/Conveyor.scala
+++ b/app/workers/Conveyor.scala
@@ -14,7 +14,7 @@ import play.api.libs.ws._
 import play.api.mvc._
 import play.api.http.HeaderNames._
 
-import models.{Channel, Hold, Item, Scheme, Subscriber, Subscription, Topic, Transfer}
+import models.{Channel, Hold, Item, Scheme, Subscriber, Subscription, Topic, TopicPick, Transfer}
 import services.Emailer
 
 /** Conveyor is the worker responsible for transmitting notifications
@@ -29,6 +29,7 @@ class ConveyorWorker extends Actor {
     case sub: Subscription => Conveyor.fulfill(sub)
     case (item: Item, subscr: Subscriber) => Conveyor.transferItem(item, subscr)
     case (hold: Hold, accept: Boolean) => Conveyor.resolveHold(hold, accept)
+    case (pick: TopicPick, accept: Boolean) => Conveyor.resolvePick(pick, accept)
     case _ => println("I'm lost")
   }
 }
@@ -72,6 +73,18 @@ object Conveyor {
     }
     // clean up hold in any case
     hold.resolve(accept)
+  }
+
+  def resolvePick(pick: TopicPick, accept: Boolean) = {
+    // keep a record as a cancelled subscription? TODO
+    if (accept) {
+      val sub = Subscriber.findById(pick.subscriberId).get
+      val topic = Topic.findById(pick.topicId).get
+      // TODO derive action from topic's scheme plan
+      fulfill(Subscription.make(pick.subscriberId, pick.topicId, "review", sub.created, new Date))
+    }
+    // clean up pick in any case
+    pick.resolve(accept)
   }
 
   private def processSubs(item: Item, subMap: Map[String, List[Subscription]]) = {

--- a/conf/evolutions/default/6.sql
+++ b/conf/evolutions/default/6.sql
@@ -1,0 +1,40 @@
+# core model types
+# --- !Ups
+
+-- Fixes #93 tag field too short: varchar(255)
+ALTER TABLE topic ALTER tag TYPE varchar;
+
+CREATE SEQUENCE agent_id_seq;
+CREATE TABLE agent (
+  id integer NOT NULL DEFAULT nextval('agent_id_seq'),
+  tag varchar UNIQUE,
+  label varchar,
+  description varchar,
+  code varchar,
+  params varchar,
+  icon varchar,
+  PRIMARY KEY(id)
+);
+
+CREATE SEQUENCE topic_pick_id_seq;
+CREATE TABLE topic_pick (
+  id integer NOT NULL DEFAULT nextval('topic_pick_id_seq'),
+  subscriber_id integer,
+  topic_id integer,
+  agent_id integer,
+  created timestamp,
+  resolved timestamp,
+  FOREIGN KEY(subscriber_id) REFERENCES subscriber(id),
+  FOREIGN KEY(topic_id) REFERENCES topic(id),
+  FOREIGN KEY(agent_id) REFERENCES agent(id),
+  PRIMARY KEY(id)
+);
+
+# --- !Downs
+
+ALTER TABLE topic ALTER tag TYPE varchar(255);
+
+DROP TABLE topic_pick;
+DROP SEQUENCE topic_pick_id_seq;
+DROP TABLE agent;
+DROP SEQUENCE agent_id_seq;

--- a/conf/routes
+++ b/conf/routes
@@ -120,6 +120,10 @@ GET      /subscriptions/browse      controllers.Application.subscriptionBrowse(f
 GET     /holds/browse               controllers.Application.holdBrowse(id: Int, page: Int ?= 0)
 GET     /hold/:id/resolve           controllers.Application.resolveHold(id: Int, accept: Boolean)
 
+# Topic Pick pages
+GET     /picks/browse               controllers.Application.pickBrowse(id: Int, page: Int ?= 0)
+GET     /pick/:id/resolve           controllers.Application.resolvePick(id: Int, accept: Boolean)
+
 # Search pages
 GET     /search                     controllers.Search.index
 GET     /search/results             controllers.Search.results(q: String, target: String ?="topic", page: Int ?=0, perpage: Int ?=25)

--- a/test/unit/AgentSpec.scala
+++ b/test/unit/AgentSpec.scala
@@ -1,0 +1,72 @@
+import org.specs2.mutable._
+import java.util.Date
+
+import play.api.test._
+import play.api.test.Helpers._
+
+import models.{Scheme, Subscriber}
+import models.{Agent, Topic, TopicPick, User}
+
+class AgentSpec extends Specification {
+
+  "Agent model" should {
+
+    "#all" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        Agent.all must haveSize(0)
+        Agent.create("tag", "label", "description", "code", "params", Some("icon"))
+        Agent.create("tag2", "label", "description", "code", "params", Some("icon"))
+        Agent.all must haveSize(2)
+        Agent.all.contains(Agent.findById(1).get) must equalTo(true)
+        Agent.all.contains(Agent.findById(2).get) must equalTo(true)
+      }
+    }
+
+    "#findById" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        Agent.all must haveSize(0)
+        Agent.create("tag", "label", "description", "code", "params", Some("icon"))
+        Agent.create("tag2", "label", "description", "code", "params", Some("icon"))
+        Agent.findById(1).get.tag must equalTo("tag")
+      }
+    }
+
+    "#findByTag" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        Agent.all must haveSize(0)
+        Agent.create("tag", "label", "description", "code", "params", Some("icon"))
+        Agent.create("tag2", "label", "description", "code", "params", Some("icon"))
+        Agent.findByTag("tag").get.id must equalTo(1)
+      }
+    }
+
+    "#create" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        Agent.all must haveSize(0)
+        Agent.create("tag", "label", "description", "code", "params", Some("icon"))
+        Agent.all must haveSize(1)
+      }
+    }
+
+    "#make" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        Agent.all must haveSize(0)
+        val a = Agent.make("tag", "label", "description", "code", "params", Some("icon"))
+        Agent.all must haveSize(1)
+        a.tag must equalTo("tag")
+      }
+    }
+
+    "#mapView" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        Agent.all must haveSize(0)
+        Agent.create("tag", "label", "description", "code", "params", Some("icon"))
+        Agent.create("tag2", "label", "description", "code", "params", Some("icon"))
+        Agent.create("tag3", "label", "description", "code", "params", Some("icon"))
+        Agent.mapView must havePair("1" -> "tag")
+        Agent.mapView must havePair("2" -> "tag2")
+        Agent.mapView must havePair("3" -> "tag3")
+      }
+    }
+  }
+}

--- a/test/unit/SubscriberSpec.scala
+++ b/test/unit/SubscriberSpec.scala
@@ -2,9 +2,7 @@ import org.specs2.mutable._
 import play.api.test._
 import play.api.test.Helpers._
 
-import models.Channel
-import models.Collection
-import models.ContentType
+import models.{Agent, Channel, Collection, ContentType}
 import models.Hold
 import models.Interest
 import models.Item
@@ -13,7 +11,7 @@ import models.ResourceMap
 import models.Scheme
 import models.Subscriber
 import models.Subscription
-import models.Topic
+import models.{Topic, TopicPick}
 import models.Transfer
 import models.User
 
@@ -419,6 +417,94 @@ class SubscriberSpec extends Specification {
 
         s.holds(0).size must equalTo(10)
         s.holds(1).size must equalTo(1)
+      }
+    }
+
+    "#pickCount" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        Subscriber.all must haveSize(0)
+        User.create("bob", "bob@example.com", "pwd", "role1")
+        Scheme.create("tag", "topic", "cat", "desc", Some("link"), Some("logo"))
+        Scheme.create("tag2", "topic", "cat", "desc", Some("link"), Some("logo"))
+        Scheme.create("tag3", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        val s = Subscriber.make(1, "Sub Name", "cat", "contact", Some("link"), Some("logo"))
+        s.pickCount must equalTo(0)
+        val top = Topic.make(1, "tag", "name")
+        val top2 = Topic.make(1, "tag2", "name")
+        val agent = Agent.make("tag", "label", "description", "code", "params", Some("icon"))
+        TopicPick.create(s.id, top.id, agent.id)
+        s.pickCount must equalTo(1)
+        TopicPick.create(s.id, top2.id, agent.id)
+        s.pickCount must equalTo(2)
+      }
+    }
+
+    "#picked" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        Subscriber.all must haveSize(0)
+        User.create("bob", "bob@example.com", "pwd", "role1")
+        Scheme.create("tag", "topic", "cat", "desc", Some("link"), Some("logo"))
+        Scheme.create("tag2", "topic", "cat", "desc", Some("link"), Some("logo"))
+        Scheme.create("tag3", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        val s = Subscriber.make(1, "Sub Name", "cat", "contact", Some("link"), Some("logo"))
+        s.holdCount must equalTo(0)
+        val top = Topic.make(1, "tag", "name")
+        val top2 = Topic.make(1, "tag2", "name")
+        val agent = Agent.make("tag", "label", "description", "code", "params", Some("icon"))
+
+        val p1 = TopicPick.make(s.id, top.id, agent.id)
+        s.pickCount must equalTo(1)
+        s.picked(top.id) must equalTo(Some(p1))
+        s.picked(top2.id) must equalTo(None)
+
+        val p2 = TopicPick.make(s.id, top2.id, agent.id)
+        s.pickCount must equalTo(2)
+        s.picked(top.id) must equalTo(Some(p1))
+        s.picked(top2.id) must equalTo(Some(p2))
+      }
+    }
+
+    "#picks" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        Subscriber.all must haveSize(0)
+        User.create("bob", "bob@example.com", "pwd", "role1")
+        Scheme.create("tag", "topic", "cat", "desc", Some("link"), Some("logo"))
+        Scheme.create("tag2", "topic", "cat", "desc", Some("link"), Some("logo"))
+        Scheme.create("tag3", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        val s = Subscriber.make(1, "Sub Name", "cat", "contact", Some("link"), Some("logo"))
+        val agent = Agent.make("tag", "label", "description", "code", "params", Some("icon"))
+        s.pickCount must equalTo(0)
+        val t = Topic.make(1, "tag", "name")
+        val t2 = Topic.make(1, "tag2", "name")
+
+        val p1 = TopicPick.make(s.id, t.id, agent.id)
+        val p2 = TopicPick.make(s.id, t2.id, agent.id)
+        s.picks(0).size must equalTo(2)
+        s.picks(0)(0) must equalTo(p2)
+        s.picks(0)(1) must equalTo(p1)
+
+        val t3 = Topic.make(1, "tag3", "name")
+        val t4 = Topic.make(1, "tag4", "name")
+        val t5 = Topic.make(1, "tag5", "name")
+        val t6 = Topic.make(1, "tag6", "name")
+        val t7 = Topic.make(1, "tag7", "name")
+        val t8 = Topic.make(1, "tag8", "name")
+        val t9 = Topic.make(1, "tag9", "name")
+        val t10 = Topic.make(1, "tag10", "name")
+        val t11 = Topic.make(1, "tag11", "name")
+
+        TopicPick.create(s.id, t3.id, agent.id)
+        TopicPick.create(s.id, t4.id, agent.id)
+        TopicPick.create(s.id, t5.id, agent.id)
+        TopicPick.create(s.id, t6.id, agent.id)
+        TopicPick.create(s.id, t7.id, agent.id)
+        TopicPick.create(s.id, t8.id, agent.id)
+        TopicPick.create(s.id, t9.id, agent.id)
+        TopicPick.create(s.id, t10.id, agent.id)
+        TopicPick.create(s.id, t11.id, agent.id)
+
+        s.picks(0).size must equalTo(10)
+        s.picks(1).size must equalTo(1)
       }
     }
 

--- a/test/unit/TopicPickSpec.scala
+++ b/test/unit/TopicPickSpec.scala
@@ -1,0 +1,97 @@
+import org.specs2.mutable._
+import java.util.Date
+
+import play.api.test._
+import play.api.test.Helpers._
+
+import models.{Scheme, Subscriber}
+import models.{Agent, Topic, TopicPick, User}
+
+class TopicPickSpec extends Specification {
+
+  "TopicPick model" should {
+
+    "#picked" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        User.create("bob1", "bob@example.com", "pwd", "role1")
+        val s = Subscriber.make(1, "Sub Name", "cat", "contact", Some("link"), Some("logo"))
+        Scheme.create("tag1", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        val top = Topic.make(1, "tag", "name")
+        val agent = Agent.make("tag", "label", "description", "code", "params", Some("icon"))
+        TopicPick.picked(top.id, s.id) must equalTo(false)
+        val p = TopicPick.make(s.id, top.id, agent.id)
+        TopicPick.picked(top.id, s.id) must equalTo(true)
+      }
+    }
+
+    "#findById" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        User.create("bob1", "bob@example.com", "pwd", "role1")
+        val s = Subscriber.make(1, "Sub Name", "cat", "contact", Some("link"), Some("logo"))
+        Scheme.create("tag1", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        val top = Topic.make(1, "tag", "name")
+        val agent = Agent.make("tag", "label", "description", "code", "params", Some("icon"))
+        TopicPick.picked(top.id, s.id) must equalTo(false)
+        val p = TopicPick.make(s.id, top.id, agent.id)
+        TopicPick.findById(1).get must equalTo(p)
+      }
+    }
+
+    "#create" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        User.create("bob1", "bob@example.com", "pwd", "role1")
+        val s = Subscriber.make(1, "Sub Name", "cat", "contact", Some("link"), Some("logo"))
+        Scheme.create("tag1", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        val top = Topic.make(1, "tag", "name")
+        val agent = Agent.make("tag", "label", "description", "code", "params", Some("icon"))
+        TopicPick.picked(top.id, s.id) must equalTo(false)
+        TopicPick.create(s.id, top.id, agent.id)
+        TopicPick.picked(top.id, s.id) must equalTo(true)
+      }
+    }
+
+    "#make" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        User.create("bob1", "bob@example.com", "pwd", "role1")
+        val s = Subscriber.make(1, "Sub Name", "cat", "contact", Some("link"), Some("logo"))
+        Scheme.create("tag1", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        val top = Topic.make(1, "tag", "name")
+        val agent = Agent.make("tag", "label", "description", "code", "params", Some("icon"))
+        TopicPick.picked(top.id, s.id) must equalTo(false)
+        val p = TopicPick.make(s.id, top.id, agent.id)
+        TopicPick.picked(top.id, s.id) must equalTo(true)
+      }
+    }
+
+    "#topic" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        User.create("bob1", "bob@example.com", "pwd", "role1")
+        val s = Subscriber.make(1, "Sub Name", "cat", "contact", Some("link"), Some("logo"))
+        Scheme.create("tag1", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        val top = Topic.make(1, "tag", "name")
+        val agent = Agent.make("tag", "label", "description", "code", "params", Some("icon"))
+        val p = TopicPick.make(s.id, top.id, agent.id)
+        p.topic must equalTo(top)
+      }
+    }
+
+    "#resolve" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        User.create("bob1", "bob@example.com", "pwd", "role1")
+        val s = Subscriber.make(1, "Sub Name", "cat", "contact", Some("link"), Some("logo"))
+        Scheme.create("tag1", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        val top = Topic.make(1, "tag", "name")
+        val agent = Agent.make("tag", "label", "description", "code", "params", Some("icon"))
+        val p = TopicPick.make(s.id, top.id, agent.id)
+        TopicPick.picked(top.id, s.id) must equalTo(true)
+        p.resolve(true)
+        TopicPick.picked(top.id, s.id) must equalTo(false)
+
+        val p2 = TopicPick.make(s.id, top.id, agent.id)
+        TopicPick.picked(top.id, s.id) must equalTo(true)
+        p2.resolve(false)
+        TopicPick.picked(top.id, s.id) must equalTo(false)
+      }
+    }
+  }
+}

